ananse/influence.py

```diff
@@ -24,6 +26,94 @@
 Expression = namedtuple("Expression", ["score", "absfc", "realfc"])
 
 
+# This piece of code is adapted from the networkx code licensed under a 3-clause license:
+# https://networkx.org/documentation/networkx-2.7/#license
+def dijkstra_prob_length(G, source, weight, cutoff=None, target=None):
+    """Uses Dijkstra's algorithm to find shortest weighted paths
+    Parameters
+    ----------
+    G : NetworkX graph
+    source : node
+        Starting node for paths.
+    weight: string
+        Name of attribute that contains weight (as a probability between 0 and 1).
+    target : node label, optional
+        Ending node for path. Search is halted when target is found.
+    cutoff : integer or float, optional
+        Minimum combined, weighted probability.
+        If cutoff is provided, only return paths with summed weight >= cutoff.
+    Returns
+    -------
+    distance : dictionary
+        A mapping from node to shortest distance to that node from one
+        of the source nodes.
+    Raises
+    ------
+    NodeNotFound
+        If the `source` is not in `G`.
+    """
+    weight_function = lambda u, v, data: -np.log10(data.get(weight, 1))  # noqa: E731
+    paths = {source: [source]}
+
+    if cutoff < 0 or cutoff > 1:
+        raise ValueError(
+            "cutoff (combined, weighted probability) should be between 0 and 1"
+        )
+    if cutoff == 0:
+        cutoff = None
+    else:
+        cutoff = -np.log10(cutoff)
+
+    G_succ = G._succ if G.is_directed() else G._adj
+
+    push = heappush
+    pop = heappop
+    dist = {}  # dictionary of final distances
+    seen = {}
+    # fringe is heapq with 3-tuples (distance,c,node)
+    # use the count c to avoid comparing nodes (may not be able to)
+    c = count()
+    fringe = []
+
+    seen[source] = 0
+    push(fringe, (0, next(c), source))
+    while fringe:
+        (d, _, v) = pop(fringe)
+        if v in dist:
+            continue  # already searched this node.
+        dist[v] = d
+        if v == target:
+            break
+        for u, e in G_succ[v].items():
+            cost = weight_function(v, u, e)
+            if cost is None:
+                continue
+
+            # This is the major difference, both probability and length are taken
+            # into account
+            length = len(paths[v])
+            vu_dist = dist[v] + cost - np.log10(1 / length)
+            if length > 1:
+                vu_dist = vu_dist + np.log10(1 / (length - 1))
```

Comment:
Could you add some comments explaining this section? In the paper we use a maximum interaction distance (max 3 steps: source TF (-> intermediate TF) -> target gene), and it's not really clear how this is enforced here. 

I think you are adding a total length cost, then reducing it by cost of previous steps?

Reply:
if the max interaction limit is important, we can limit the graph to nodes within range with 
```python
sub_grn = nx.generators.ego_graph(grn, node, distance=2)
paths, weights = dijkstra_prob_length(sub_grn, node, "weight", cutoff=0.6)
```